src/zarr/codecs/mixins.py

```diff
@@ -0,0 +1,183 @@
+    ArrayArrayCodec,
+    ArrayBytesCodec,
+    ArrayBytesCodecPartialDecodeMixin,
+    ArrayBytesCodecPartialEncodeMixin,
+    ByteGetter,
+    ByteSetter,
+    BytesBytesCodec,
+)
+from zarr.common import ArraySpec, BytesLike, SliceSelection, concurrent_map
+from zarr.config import config
+
+
+CodecInput = TypeVar("CodecInput", bound=np.ndarray | BytesLike)
+CodecOutput = TypeVar("CodecOutput", bound=np.ndarray | BytesLike)
+
+
+async def batching_helper(
+    func: Callable[[CodecInput, ArraySpec], Awaitable[CodecOutput | None]],
+    batch_info: Iterable[tuple[CodecInput | None, ArraySpec]],
+) -> list[CodecOutput | None]:
+    return await concurrent_map(
+        [(chunk_array, chunk_spec) for chunk_array, chunk_spec in batch_info],
+        noop_for_none(func),
+        config.get("async.concurrency"),
+    )
+
+
+def noop_for_none(
+    func: Callable[[CodecInput, ArraySpec], Awaitable[CodecOutput | None]],
+) -> Callable[[CodecInput | None, ArraySpec], Awaitable[CodecOutput | None]]:
+    async def wrap(chunk: CodecInput | None, chunk_spec: ArraySpec) -> CodecOutput | None:
+        if chunk is None:
+            return None
+        return await func(chunk, chunk_spec)
+
+    return wrap
+
+
+class ArrayArrayCodecBatchMixin(ArrayArrayCodec):
+    @abstractmethod
+    async def decode_single(
+        self,
+        chunk_array: np.ndarray,
+        chunk_spec: ArraySpec,
+    ) -> np.ndarray:
+        pass
+
+    async def decode(
+        self,
+        chunk_arrays_and_specs: Iterable[tuple[np.ndarray | None, ArraySpec]],
+    ) -> Iterable[np.ndarray | None]:
+        return await batching_helper(self.decode_single, chunk_arrays_and_specs)
+
+    @abstractmethod
+    async def encode_single(
+        self,
+        chunk_array: np.ndarray,
+        chunk_spec: ArraySpec,
+    ) -> np.ndarray | None:
+        pass
+
+    async def encode(
+        self,
+        chunk_arrays_and_specs: Iterable[tuple[np.ndarray | None, ArraySpec]],
+    ) -> Iterable[np.ndarray | None]:
+        return await batching_helper(self.encode_single, chunk_arrays_and_specs)
+
+
+class ArrayBytesCodecBatchMixin(ArrayBytesCodec):
+    @abstractmethod
+    async def decode_single(
+        self,
+        chunk_bytes: BytesLike,
+        chunk_spec: ArraySpec,
+    ) -> np.ndarray:
+        pass
+
+    async def decode(
+        self,
+        chunk_bytes_and_specs: Iterable[tuple[BytesLike | None, ArraySpec]],
+    ) -> Iterable[np.ndarray | None]:
+        return await batching_helper(self.decode_single, chunk_bytes_and_specs)
+
+    @abstractmethod
+    async def encode_single(
+        self,
+        chunk_array: np.ndarray,
+        chunk_spec: ArraySpec,
+    ) -> BytesLike | None:
+        pass
+
+    async def encode(
+        self,
+        chunk_arrays_and_specs: Iterable[tuple[np.ndarray | None, ArraySpec]],
+    ) -> Iterable[BytesLike | None]:
+        return await batching_helper(self.encode_single, chunk_arrays_and_specs)
+
```

Comment:
The amount of logic duplicated here really stands out to me. I wonder if there is an abstraction we are missing that could simplify this

Reply:
Can you elaborate? The logic is reused through the `batching_helper`. Large parts of the signature are duplicated but differ in details due to the nature of the 3 separate codec classes. What abstraction would you propose?

Reply:
I was able to get rid of a lot of this code by using `typing.Generic`.